apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerParameterCollection.cs

```diff
@@ -0,0 +1,196 @@
+        }
+
+        /// <inheritdoc />
+        public override void CopyTo(Array array, int index)
+        {
+            foreach (var item in InnerList)
+            {
+                array?.SetValue(item, index);
+                index++;
+            }
+        }
+
+        /// <inheritdoc />
+        public override IEnumerator GetEnumerator()
+        {
+            return InnerList.GetEnumerator();
+        }
+
+        /// <inheritdoc />
+        public override int IndexOf(object value)
+        {
+            return InnerList.IndexOf(value as SpannerParameter);
+        }
+
+        /// <inheritdoc />
+        public override int IndexOf(string parameterName)
+        {
+            return InnerList.FindIndex(x => x.ParameterName.Equals(parameterName));
+        }
+
+        /// <inheritdoc />
+        public override void Insert(int index, object value)
+        {
+            InnerList.Insert(index, (SpannerParameter) value);
+        }
+
+        /// <inheritdoc />
+        public override void Remove(object value)
+        {
+            InnerList.Remove((SpannerParameter) value);
+        }
+
+        /// <inheritdoc />
+        public override void RemoveAt(int index)
+        {
+            InnerList.RemoveAt(index);
+        }
+
+        /// <inheritdoc />
+        public override void RemoveAt(string parameterName)
+        {
+            InnerList.RemoveAt(IndexOf(parameterName));
+        }
+
+        /// <inheritdoc />
+        protected override DbParameter GetParameter(int index)
+        {
+            return InnerList[index];
+        }
+
+        /// <inheritdoc />
+        protected override DbParameter GetParameter(string parameterName)
+        {
+            var index = IndexOf(parameterName);
+            if (index == -1)
+                return null;
+            return InnerList[index];
+        }
+
+        /// <inheritdoc />
+        protected override void SetParameter(int index, DbParameter value)
+        {
+            InnerList[index] = (SpannerParameter) value;
+        }
+
+        /// <inheritdoc />
+        protected override void SetParameter(string parameterName, DbParameter value)
+        {
+            var index = IndexOf(parameterName);
+            if (index == -1)
+                InnerList.Add((SpannerParameter) value);
+            else
+                InnerList[index] = (SpannerParameter) value;
+        }
+
+        internal void FillSpannerInternalValues(MapField<string, Value> valueDictionary)
+        {
+            foreach (var parameter in InnerList)
+            {
+                valueDictionary[parameter.ParameterName] = TypeMap.ToValue(parameter.Value, parameter.TypeCode);
+            }
+        }
+
+        internal void FillSpannerInternalTypes(MapField<string, V1.Type> typeDictionary)
+        {
+            foreach (var parameter in InnerList)
+            {
+                typeDictionary[parameter.ParameterName] = new V1.Type { Code = parameter.TypeCode };
```

Comment:
Do we ever modify these? If it's all internal, I was wondering whether it would be worth having a cache of Type objects (if they're really just the type code).